sr/robot/astoria.py

```diff
@@ -0,0 +1,294 @@
+    """
+
+    name: ClassVar[str] = "start_button"
+
+    event_name: str
+    sender_name: str
+    priority: int = 0
+
+    def __gt__(self, other: "StartButtonBroadcastEvent") -> bool:
+        return self.priority > other.priority
+
+
+# Module Logic
+
+class AstoriaInterface:
+    def __init__(self, mqtt_client: MQTTClient) -> None:
+        self._mqtt = mqtt_client
+        self._metadata_lock = Lock()
+        self._metadata = Metadata()
+        self._mount_path = Path("/dev/null")
+        self._mount_path_lock = Lock()
+        self._start_pressed = Event()
+
+        self._mqtt.subscribe("broadcast/start_button", self._process_remote_start)
+        self._mqtt.subscribe("astmetad", self._process_metadata_update)
+        self._mqtt.subscribe("astprocd", self._handle_astprocd_message)
+
+    def _process_remote_start(
+        self, client: MQTT, userdata: Any, msg: MQTTMessage,
+    ) -> None:
+        """Receive a remote start event."""
+        try:
+            event: StartButtonBroadcastEvent = StartButtonBroadcastEvent.parse_raw(msg.payload)
+            LOGGER.debug(f"Received a {event.event_name} event from {event.sender_name}")
+            self._start_pressed.set()
+        except (JSONDecodeError, ValidationError) as e:
+            LOGGER.debug(f"Unable to parse start button message: {e}")
+
+    def _process_metadata_update(
+        self, client: MQTT, userdata: Any, msg: MQTTMessage,
+    ) -> None:
+        """Update the metadata."""
+        try:
+            MetadataManagerMessage.update_forward_refs()
+            message = MetadataManagerMessage.parse_raw(msg.payload)
+            if message.status == MetadataManagerMessage.Status.RUNNING:
+                LOGGER.debug("Received metadata")
+                with self._metadata_lock:
+                    self._metadata = message.metadata
+            else:
+                LOGGER.debug("Cannot get metadata, astmetad is not running")
+        except (JSONDecodeError, ValidationError) as e:
+            LOGGER.debug(f"Unable to parse metadata message: {e}")
+
+    def _handle_astprocd_message(
+        self, client: MQTT, userdata: Any, msg: MQTTMessage,
+    ) -> None:
+        """Update the mount path."""
+        try:
+            message = ProcessManagerMessage.parse_raw(msg.payload)
+            if message.status == ProcessManagerMessage.Status.RUNNING:
+                LOGGER.debug("Received process info")
+                if message.disk_info:
+                    with self._mount_path_lock:
+                        self._mount_path = message.disk_info.mount_path
+            else:
+                LOGGER.debug("Cannot get process info, astprocd is not running")
+        except (JSONDecodeError, ValidationError) as e:
+            LOGGER.debug(f"Unable to parse process message: {e}")
+
+    def fetch_current_metadata(self) -> Metadata:
+        """Fetch the current metadata."""
+        with self._metadata_lock:
+            return self._metadata
```

Comment:
It does make me a bit uncomfortable that we're doing thread safe protection but we're returning this mutable thing. How would you feel about a `deepcopy` here?

Reply:
28f4c9008c93f1664160a7fb1843a9786f41b83b